src/osd/sdl/man/mame.6
.\"  -*- nroff -*-
.\"
.\" mame.6
.\"
.\" Man page created from usage and source information:
.\" * commands: see src/emu/clifront.c clifront.h
.\" * options: core entries, see src/emu/emuopts.c emuopts.h
.\"            SDL-specific entries, see src/osd/sdl/sdlmain.c osdsdl.h
.\" Cesare Falco <cesare.falco@gmail.com>, February 2007
.\"
.\" Also, some text borrowed from the xmame 0.106 man page,
.\" done by Rene Herrmann <rene.herrmann@berlin.de>, September 2000
.\" and updated by Andrew Burton <burtona@gol.com>, July 2003
.\"
.\"
.TH MAME 1 2012-03-25 0.145u5 "MAME \- The Multiple Arcade Machine Emulator"
.\"
.\"
.\" NAME chapter
.SH NAME
MAME \- The Multiple Arcade Machine Emulator
.\"
.\"
.\" SYNOPSIS chapter
.SH SYNOPSIS
.B mame
.RI [ options ]
.I gamename
.\"
.\"
.\" DESCRIPTION chapter
.SH DESCRIPTION
Started in 1997 by Nicola Salmoria, MAME was originally intended as a series
of emulators for individual games, which were later combined into a single
multi\-game emulator. In the following years, MAME grew over and over up to
the actual size, with more than 100 contributors to the project.
.\"
.\"
.\" OPTIONS chapter
.SH OPTIONS
.\"
.\" *******************************************************
.SS Core commands
.\" *******************************************************
.TP
.B \-help, \-?
Displays current MAME version and copyright notice.
.TP
.B \-validate, \-valid
Performs internal validation on every driver in the system. Run this
before submitting changes to ensure that you haven't violated any of
the core system rules.
.\"
.\" *******************************************************
.SS Configuration commands
.\" *******************************************************
.TP
.B \-createconfig, \-cc
Creates the default \fBmame.ini\fP file in the current directory. All the
configuration options (not commands) described below can be permanently
changed by editing this configuration file.
.TP
.B \-showconfig, \-sc
Displays the current configuration settings.
.TP
.B \-showusage, \-su
Displays a summary of all the command line options. For options that
are not mentioned here, the short summary given by
\fBmame \-showusage\fR is usually sufficient.
.\"
.\" *******************************************************
.SS Frontend commands
.\" *******************************************************
.B NOTE:
By default, all the '\-list' commands below write info to the screen.
If you wish to write the info to a textfile instead, use redirection, e.g.
.B mame \-listxml > ~/mamelist.xml
writes the full list of supported game to file \fImamelist.xml\fR in your home
directory.
.TP
.B \-listxml, \-lx \fR[\fIgamename\fR|\fIwildcard\fR]
List comprehensive details for all of the supported games. The output
is quite long, so it is usually better to redirect this into a file.
The output is in XML format. By default all games are listed; however,
you can limit this list by specifying a driver name or wildcard after
the \-listxml command.
.TP
.B \-listfull, \-ll \fR[\fIgamename\fR|\fIwildcard\fR]
Displays a list of game driver names and descriptions. By default all
games are listed; however, you can limit this list by specifying a
driver name or wildcard after the \-listfull command.
.TP
.B \-listsource, \-ls \fR[\fIgamename\fR|\fIwildcard\fR]
Displays a list of drivers and the names of the source files their game
drivers live in. Useful for finding which driver a game runs on in
order to fix bugs. By default all games are listed; however, you can
limit this list by specifying a driver name or wildcard after the
\-listsource command.
.TP
.B \-listclones, \-lc \fR[\fIgamename\fR|\fIwildcard\fR]
Displays a list of clones. By default all clones are listed; however,
you can limit this list by specifying a driver name or wildcard after
the \-listclones command.
.TP
.B \-listbrothers, \-lb \fR[\fIgamename\fR|\fIwildcard\fR]
Displays a list of "brothers" or other drivers from same sourcefile.
By default all games are listed; however, you can limit this list by
specifying a driver name or wildcard after the \-listbrothers command.
.TP
.B \-listcrc
Displays a full list of CRCs of all ROM images referenced by all
drivers within MAME code.
.TP
.B \-listroms \fIgamename
Displays a list of ROM images referenced by the specified game.
.TP
.B \-listsamples \fIgamename
Displays a list of samples referenced by the specified game.
.TP
.B \-verifyroms \fR[\fIgamename\fR|\fIwildcard\fR]
Checks for invalid or missing ROM images. By default all drivers that
have valid ZIP files or directories in the rompath are verified;
however, you can limit this list by specifying a driver name or
wildcard after the \-verifyroms command.
.TP
.B \-verifysamples \fR[\fIgamename\fR|\fIwildcard\fR]
Checks for invalid or missing samples. By default all drivers that
have valid ZIP files or directories in the samplepath are verified;
however, you can limit this list by specifying a driver name or
wildcard after the \-verifyroms command.
.TP
.B \-romident
Attempts to identify ROM files, if they are known to MAME, in the
specified .zip file or directory. This command can be used to try and
identify ROM sets taken from unknown boards. On exit, the errorlevel
is returned as one of the following:
.br
\fB0\fR  all files were identified
.br
\fB7\fR  all files were identified except for some "non\-ROM" files
.br
\fB8\fR  some files were identified
.br
\fB9\fR  no files were identified
.TP
.B \-listdevices, \-ld
Output the list of devices referenced by a given game or set of games.
.TP
.B \-listslots, \-lslot
Output the list of available slots and slot devices for the system.
.TP
.B \-listmedia, \-lm
Output the list of available media for the system.
.TP
.B \-listsoftware, \-lsoft
Output the list of known software for the system.
.TP
.B \-getsoftlist, \-glist
Retrieve software list by name.
.\"
.\" *******************************************************
.SS Configuration options
.\" *******************************************************
.TP
.B \-[no]readconfig, \-[no]rc
Enables or disables the reading of the config files. When enabled
(which is the default), MAME reads the following config files in order:
.br
1. \fBmame.ini\fR
.br
the main configuration file
.br
2. \fI[name]\fB.ini\fR
.br
where \fIname\fR is your executable name, i.e. mame unless you changed it
(e.g. if you renamed mame to mame0137, the parsed file will be
\fImame0137.ini\fR)
.br
3. \fBdebug.ini\fR, if the debugger is enabled
.br
4. \fBvector.ini\fR, for vector games only
.br
5. \fI[driver]\fB.ini\fR
.br
based on the source filename of the game driver
.br
6. \fI[parent]\fB.ini\fR
.br
for clones only, may be called recursively
.br
7. \fI[gamename]\fB.ini\fR
.br
note this sometimes resolves to the same of the source driver
.br
The settings in the later ini's override those in the earlier ini's.
So, for example, if you wanted to disable overlay effects in the
vector games, you can create a vector.ini with the "effect none" line
in it, and it will override whatever effect value you have in your
mame.ini. The default is ON (\-readconfig).
.TP
.B \-writeconfig, \-wc
Write configuration to \fI[driver]\fR.ini on exit.
.\"
.\" *******************************************************
.SS Search path options
.\" *******************************************************
.B IMPORTANT\fR: Please use the path, directory and file options in
mame.ini \fBONLY\fR. Otherwise, the outcome may be unpredictable and not
consistent across releases.
.TP
.B \-rompath, \-rp, \-biospath, \-bp, \fIpathname
Specifies a list of paths within which to find ROM or hard disk images.
Multiple paths can be specified by separating them with semicolons.
The default is 'roms' (that is, a directory "roms" in the same directory
as the MAME executable).
.TP
.B \-samplepath, \-sp \fIpathname
Specifies a list of paths within which to find sample files. Multiple
paths can be specified by separating them with semicolons. The default
is 'samples' (that is, a directory "samples" in the same directory as
the MAME executable).
.TP
.B \-artpath, \-artwork_directory \fIpathname
Specifies a list of paths within which to find artwork files. Multiple
paths can be specified by separating them with semicolons. The default
is 'artwork' (that is, a directory "artwork" in the same directory as
the MAME executable).
.TP
.B \-ctrlrpath, \-ctrlr_directory \fIpathname
Specifies a list of paths within which to find controller\-specific
configuration files. Multiple paths can be specified by separating
them with semicolons. The default is 'ctrlr' (that is, a directory
"ctrlr" in the same directory as the MAME executable).
.TP
.B \-inipath \fIpathname
Specifies a list of paths within which to find .INI files. Multiple
paths can be specified by separating them with semicolons. The default
is '/etc/mame/'.
.TP
.B \-fontpath \fIpathname
Specifies a list of paths within which to find .BDF font files. Multiple
paths can be specified by separating them with semicolons. The default
is '.' (that is, search in the same directory as the MAME executable).
.TP
.B \-cheatpath \fIpathname
Specifies a list of paths within which to find cheat files. Multiple
paths can be specified by separating them with semicolons. The default
is 'cheat' (that is, a directory "cheat" in the same directory as
the MAME executable).
.TP
.B \-crosshairpath \fIpathname
Specifies a list of paths within which to find crosshair files. Multiple
paths can be specified by separating them with semicolons. The default
is 'crosshair' (that is, a directory "crosshair" in the same directory as
the MAME executable).  If the Crosshair is set to default in the menu,
MAME will look for gamename\cross#.png and then cross#.png in the
specified crosshairpath, where # is the player number.  Failing that,
MAME will use built\-in default crosshairs.
.\"
.\" *******************************************************
.SS Output Directory Options
.\" *******************************************************
.TP
.B \-cfg_directory \fIpathname
Specifies a single directory where configuration files are stored.
Configuration files store user configurable settings that are read at
startup and written when MAME exits. The default is 'cfg' (that is,
a directory "cfg" in the same directory as the MAME executable). If this
directory does not exist, it will be automatically created.
.TP
.B \-nvram_directory \fIpathname
Specifies a single directory where NVRAM files are stored. NVRAM files
store the contents of EEPROM and non\-volatile RAM (NVRAM) for games
which used this type of hardware. This data is read at startup and
written when MAME exits. The default is 'nvram' (that is, a directory
"nvram" in the same directory as the MAME executable). If this directory
does not exist, it will be automatically created.
.TP
.B \-memcard_directory \fIpathname
Specifies a single directory where memory card files are stored. Memory
card files store the contents of removable memory cards for games which
used this type of hardware. This data is read and written under control
of the user via the "Memory Card" menu in the user interface. The
default is 'memcard' (that is, a directory "memcard" in the same
directory as the MAME executable). If this directory does not exist,
it will be automatically created.
.TP
.B \-input_directory \fIpathname
Specifies a single directory where input recording files are stored.
Input recordings are created via the \-record option and played back via
the \-playback option. The default is 'inp' (that is, a directory
"inp" in the same directory as the MAME executable). If this directory
does not exist, it will be automatically created.
.TP
.B \-state_directory \fIpathname
Specifies a single directory where save state files are stored. Save
state files are read and written either upon user request, or when using
the \-autosave option. The default is 'sta' (that is, a directory "sta"
in the same directory as the MAME executable). If this directory does
not exist, it will be automatically created.
.TP
.B \-snapshot_directory \fIpathname
Specifies a single directory where screen snapshots are stored, when
requested by the user. The default is 'snap' (that is, a directory
"snap" in the same directory as the MAME executable). If this directory
does not exist, it will be automatically created.
.TP
.B \-diff_directory \fIpathname
Specifies a single directory where hard drive differencing files are
stored. Hard drive differencing files store any data that is written
back to a hard disk image, in order to preserve the original image. The
differencing files are created at startup when a game with a hard disk
image. The default is 'diff' (that is, a directory "diff" in the same
directory as the MAME executable). If this directory does not exist,
it will be automatically created.
.TP
.B \-comment_directory \fIpathname
Specifies a single directory where debugger comment files are stored.
Debugger comment files are written by the debugger when comments are
added to the disassembly for a game. The default is 'comments' (that is,
a directory "comments" in the same directory as the MAME executable).
If this directory does not exist, it will be automatically created.
.\"
.\" *******************************************************
.SS State/playback options
.\" *******************************************************
.TP
.B \-state \fIslot
Immediately after starting the specified game, will cause the save
state in the specified \fIslot\fP to be loaded.
.TP
.B \-[no]autosave
When enabled, automatically creates a save state file when exiting MAME
and automatically attempts to reload it when later starting MAME with
the same game. This only works for games that have explicitly enabled
save state support in their driver. The default is OFF (\-noautosave).
.TP
.B \-playback, \-pb \fIfilename
Specifies a file from which to play back a series of game inputs. This
feature does not work reliably for all games, but can be used to watch
a previously recorded game session from start to finish. In order to
make things consistent, you should only record and playback with all
configuration (.cfg), NVRAM (.nv), and memory card files deleted. The
default is NULL (no playback).
.TP
.B \-record, \-rec \fIfilename
Specifies a file to record all input from a game session. This can be
used to record a game session for later playback. This feature does not
work reliably for all games, but can be used to watch a previously
recorded game session from start to finish. In order to make things
consistent, you should only record and playback with all configuration
(.cfg), NVRAM (.nv), and memory card files deleted. The default is NULL
(no recording).
.TP
.B \-snapname \fIname
Describes how MAME should name files for snapshots. \fIname\fP is a string
that provides a template that is used to generate a filename. Three
simple substitutions are provided: the / character represents the
path separator on any target platform (even Windows); the string \fI%g\fP
represents the driver name of the current game; and the string \fI%i\fP
represents an incrementing index. If \fI%i\fP is omitted, then each
snapshot taken will overwrite the previous one; otherwise, MAME will
find the next empty value for \fI%i\fP and use that for a filename. The
default is \fI%g/%i\fP, which creates a separate folder for each game,
and names the snapshots under it starting with 0000 and increasing
from there.
.TP
.B \-snapsize \fIwidth\fPx\fIheight\fP
Hard\-codes the size for snapshots and movie recording. By default,
MAME will create snapshots at the game's current resolution in raw
pixels, and will create movies at the game's starting resolution in
raw pixels. If you specify this option, then MAME will create both
snapshots and movies at the size specified, and will bilinear filter
the result. Note that this size does not automatically rotate if the 
game is vertically oriented. The default is 'auto'.
.TP
.B \-snapview \fIviewname\fP
Specifies the view to use when rendering snapshots and movies. By
default, both use a special 'internal' view, which renders a separate
snapshot per screen or renders movies only of the first screen. By
specifying this option, you can override this default behavior and
select a single view that will apply to all snapshots and movies.
Note that \fIviewname\fP does not need to be a perfect match; rather, it
will select the first view whose name matches all the characters
specified by \fIviewname\fP. For example, \-snapview native will match the 
"Native (15:14)" view even though it is not a perfect match. 
\fIviewname\fP can also be 'auto', which selects the first view with all

src/emu/mame.c
/***************************************************************************

    mame.c

    Controls execution of the core MAME system.

    Copyright Nicola Salmoria and the MAME Team.
    Visit http://mamedev.org for licensing and usage restrictions.

****************************************************************************

    Since there has been confusion in the past over the order of
    initialization and other such things, here it is, all spelled out
    as of January, 2008:

    main()
        - does platform-specific init
        - calls mame_execute() [mame.c]

        mame_execute() [mame.c]
            - calls mame_validitychecks() [validity.c] to perform validity checks on all compiled drivers
            - begins resource tracking (level 1)
            - calls create_machine [mame.c] to initialize the running_machine structure
            - calls init_machine() [mame.c]

            init_machine() [mame.c]
                - calls fileio_init() [fileio.c] to initialize file I/O info
                - calls config_init() [config.c] to initialize configuration system
                - calls input_init() [input.c] to initialize the input system
                - calls output_init() [output.c] to initialize the output system
                - calls state_init() [state.c] to initialize save state system
                - calls state_save_allow_registration() [state.c] to allow registrations
                - calls palette_init() [palette.c] to initialize palette system
                - calls render_init() [render.c] to initialize the rendering system
                - calls ui_init() [ui.c] to initialize the user interface
                - calls generic_machine_init() [machine/generic.c] to initialize generic machine structures
                - calls generic_video_init() [video/generic.c] to initialize generic video structures
                - calls generic_sound_init() [audio/generic.c] to initialize generic sound structures
                - calls timer_init() [timer.c] to reset the timer system
                - calls osd_init() [osdepend.h] to do platform-specific initialization
                - calls input_port_init() [inptport.c] to set up the input ports
                - calls rom_init() [romload.c] to load the game's ROMs
                - calls memory_init() [memory.c] to process the game's memory maps
                - calls watchdog_init() [watchdog.c] to initialize the watchdog system
                - calls the driver's DRIVER_INIT callback
                - calls device_list_start() [devintrf.c] to start any devices
                - calls video_init() [video.c] to start the video system
                - calls tilemap_init() [tilemap.c] to start the tilemap system
                - calls crosshair_init() [crsshair.c] to configure the crosshairs
                - calls sound_init() [sound.c] to start the audio system
                - calls debugger_init() [debugger.c] to set up the debugger
                - calls the driver's MACHINE_START, SOUND_START, and VIDEO_START callbacks
                - calls cheat_init() [cheat.c] to initialize the cheat system
                - calls image_init() [image.c] to initialize the image system

            - calls config_load_settings() [config.c] to load the configuration file
            - calls nvram_load [machine/generic.c] to load NVRAM
            - calls ui_display_startup_screens() [ui.c] to display the the startup screens
            - begins resource tracking (level 2)
            - calls soft_reset() [mame.c] to reset all systems

                -------------------( at this point, we're up and running )----------------------

            - calls scheduler->timeslice() [schedule.c] over and over until we exit
            - ends resource tracking (level 2), freeing all auto_mallocs and timers
            - calls the nvram_save() [machine/generic.c] to save NVRAM
            - calls config_save_settings() [config.c] to save the game's configuration
            - calls all registered exit routines [mame.c]
            - ends resource tracking (level 1), freeing all auto_mallocs and timers

        - exits the program

***************************************************************************/

#include "emu.h"
#include "emuopts.h"
#include "osdepend.h"
#include "config.h"
#include "debugger.h"
#include "image.h"
#include "profiler.h"
#include "render.h"
#include "cheat.h"
#include "ui.h"
#include "uimenu.h"
#include "uiinput.h"
#include "streams.h"
#include "crsshair.h"
#include "validity.h"
#include "debug/debugcon.h"

#include <time.h>



/***************************************************************************
    GLOBAL VARIABLES
***************************************************************************/

/* the current options */
static core_options *mame_opts;

/* started empty? */
static bool started_empty;

static running_machine *global_machine;

/* output channels */
static output_callback_func output_cb[OUTPUT_CHANNEL_COUNT];
static void *output_cb_param[OUTPUT_CHANNEL_COUNT];

/* the "disclaimer" that should be printed when run with no parameters */
const char mame_disclaimer[] =
	"MAME is an emulator: it reproduces, more or less faithfully, the behaviour of\n"
	"several arcade machines. But hardware is useless without software, so an image\n"
	"of the ROMs which run on that hardware is required. Such ROMs, like any other\n"
	"commercial software, are copyrighted material and it is therefore illegal to\n"
	"use them if you don't own the original arcade machine. Needless to say, ROMs\n"
	"are not distributed together with MAME. Distribution of MAME together with ROM\n"
	"images is a violation of copyright law and should be promptly reported to the\n"
	"authors so that appropriate legal action can be taken.\n";



/***************************************************************************
    FUNCTION PROTOTYPES
***************************************************************************/

static int parse_ini_file(core_options *options, const char *name, int priority);



/***************************************************************************
    CORE IMPLEMENTATION
***************************************************************************/

/*-------------------------------------------------
    mame_is_valid_machine - return true if the
    given machine is valid
-------------------------------------------------*/

int mame_is_valid_machine(running_machine *machine)
{
	return (machine != NULL && machine == global_machine);
}


/*-------------------------------------------------
    mame_execute - run the core emulation
-------------------------------------------------*/

int mame_execute(core_options *options)
{
	bool firstgame = true;
	bool firstrun = true;

	// loop across multiple hard resets
	bool exit_pending = false;
	int error = MAMERR_NONE;
	while (error == MAMERR_NONE && !exit_pending)
	{
		// specify the global mame_options
		mame_opts = options;

		// convert the specified gamename to a driver
		astring gamename;
		core_filename_extract_base(&gamename, options_get_string(options, OPTION_GAMENAME), true);
		const game_driver *driver = driver_get_name(gamename);

		// if no driver, use the internal empty driver
		if (driver == NULL)
		{
			driver = &GAME_NAME(empty);
			if (firstgame)
				started_empty = true;
		}

		// otherwise, perform validity checks before anything else
		else if (mame_validitychecks(driver) != 0)
			return MAMERR_FAILED_VALIDITY;

		firstgame = false;

		// parse any INI files as the first thing
		if (options_get_bool(options, OPTION_READCONFIG))
		{
			options_revert(options, OPTION_PRIORITY_INI);
			mame_parse_ini_files(options, driver);
		}

		// create the machine configuration
		const machine_config *config = global_alloc(machine_config(driver->machine_config));

		// create the machine structure and driver
		running_machine *machine = global_alloc(running_machine(*driver, *config, *options, started_empty));

		// looooong term: remove this
		global_machine = machine;

		// run the machine
		error = machine->run(firstrun);
		firstrun = false;

		// check the state of the machine
		if (machine->new_driver_pending())
		{
			options_set_string(options, OPTION_GAMENAME, machine->new_driver_name(), OPTION_PRIORITY_CMDLINE);
			firstrun = true;
		}
		if (machine->exit_pending())
			exit_pending = true;

		// destroy the machine and the config
		global_free(machine);
		global_free(config);
		global_machine = NULL;

		// reset the options
		mame_opts = NULL;
	}

	// return an error
	return error;
}


/*-------------------------------------------------
    mame_options - accesses the options for the
    currently running emulation
-------------------------------------------------*/

core_options *mame_options(void)
{
	assert(mame_opts != NULL);
	return mame_opts;
}



/***************************************************************************
    OUTPUT MANAGEMENT
***************************************************************************/

/*-------------------------------------------------
    mame_set_output_channel - configure an output
    channel
-------------------------------------------------*/

void mame_set_output_channel(output_channel channel, output_callback_func callback, void *param, output_callback_func *prevcb, void **prevparam)
{
	assert(channel < OUTPUT_CHANNEL_COUNT);
	assert(callback != NULL);

	/* return the originals if requested */
	if (prevcb != NULL)
		*prevcb = output_cb[channel];
	if (prevparam != NULL)
		*prevparam = output_cb_param[channel];

	/* set the new ones */
	output_cb[channel] = callback;
	output_cb_param[channel] = param;
}


/*-------------------------------------------------
    mame_file_output_callback - default callback
    for file output
-------------------------------------------------*/

void mame_file_output_callback(void *param, const char *format, va_list argptr)
{
	vfprintf((FILE *)param, format, argptr);
}


/*-------------------------------------------------
    mame_null_output_callback - default callback
    for no output
-------------------------------------------------*/

void mame_null_output_callback(void *param, const char *format, va_list argptr)
{
}


/*-------------------------------------------------
    mame_printf_error - output an error to the
    appropriate callback
-------------------------------------------------*/

void mame_printf_error(const char *format, ...)
{
	va_list argptr;

	/* by default, we go to stderr */
	if (output_cb[OUTPUT_CHANNEL_ERROR] == NULL)
	{
		output_cb[OUTPUT_CHANNEL_ERROR] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_ERROR] = stderr;
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_ERROR])(output_cb_param[OUTPUT_CHANNEL_ERROR], format, argptr);
	va_end(argptr);
}


/*-------------------------------------------------
    mame_printf_warning - output a warning to the
    appropriate callback
-------------------------------------------------*/

void mame_printf_warning(const char *format, ...)
{
	va_list argptr;

	/* by default, we go to stderr */
	if (output_cb[OUTPUT_CHANNEL_WARNING] == NULL)
	{
		output_cb[OUTPUT_CHANNEL_WARNING] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_WARNING] = stderr;
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_WARNING])(output_cb_param[OUTPUT_CHANNEL_WARNING], format, argptr);
	va_end(argptr);
}


/*-------------------------------------------------
    mame_printf_info - output info text to the
    appropriate callback
-------------------------------------------------*/

void mame_printf_info(const char *format, ...)
{
	va_list argptr;

	/* by default, we go to stdout */
	if (output_cb[OUTPUT_CHANNEL_INFO] == NULL)
	{
		output_cb[OUTPUT_CHANNEL_INFO] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_INFO] = stdout;
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_INFO])(output_cb_param[OUTPUT_CHANNEL_INFO], format, argptr);
	va_end(argptr);
}


/*-------------------------------------------------
    mame_printf_verbose - output verbose text to
    the appropriate callback
-------------------------------------------------*/

void mame_printf_verbose(const char *format, ...)
{
	va_list argptr;

	/* if we're not verbose, skip it */
	if (mame_opts == NULL || !options_get_bool(mame_options(), OPTION_VERBOSE))
		return;

	/* by default, we go to stdout */
	if (output_cb[OUTPUT_CHANNEL_VERBOSE] == NULL)
	{
		output_cb[OUTPUT_CHANNEL_VERBOSE] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_VERBOSE] = stdout;
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_VERBOSE])(output_cb_param[OUTPUT_CHANNEL_VERBOSE], format, argptr);
	va_end(argptr);
}


/*-------------------------------------------------
    mame_printf_debug - output debug text to the
    appropriate callback
-------------------------------------------------*/

void mame_printf_debug(const char *format, ...)
{
	va_list argptr;

	/* by default, we go to stderr */
	if (output_cb[OUTPUT_CHANNEL_DEBUG] == NULL)
	{
#ifdef MAME_DEBUG
		output_cb[OUTPUT_CHANNEL_DEBUG] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_DEBUG] = stdout;
#else
		output_cb[OUTPUT_CHANNEL_DEBUG] = mame_null_output_callback;
		output_cb_param[OUTPUT_CHANNEL_DEBUG] = NULL;
#endif
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_DEBUG])(output_cb_param[OUTPUT_CHANNEL_DEBUG], format, argptr);
	va_end(argptr);
}


/*-------------------------------------------------
    mame_printf_log - output log text to the
    appropriate callback
-------------------------------------------------*/

#ifdef UNUSED_FUNCTION
void mame_printf_log(const char *format, ...)
{
	va_list argptr;

	/* by default, we go to stderr */
	if (output_cb[OUTPUT_CHANNEL_LOG] == NULL)
	{
		output_cb[OUTPUT_CHANNEL_LOG] = mame_file_output_callback;
		output_cb_param[OUTPUT_CHANNEL_LOG] = stderr;
	}

	/* do the output */
	va_start(argptr, format);
	(*output_cb[OUTPUT_CHANNEL_LOG])(output_cb_param[OUTPUT_CHANNEL_LOG], format, argptr);
	va_end(argptr);
}
#endif


/***************************************************************************
    MISCELLANEOUS
***************************************************************************/

/*-------------------------------------------------
    popmessage - pop up a user-visible message
-------------------------------------------------*/

void CLIB_DECL popmessage(const char *format, ...)
{
	// if the format is NULL, it is a signal to clear the popmessage
	if (format == NULL)
		ui_popup_time(0, " ");

	// otherwise, generate the buffer and call the UI to display the message
	else
	{
		astring temp;
		va_list arg;

		// dump to the buffer
		va_start(arg, format);
		temp.vprintf(format, arg);
		va_end(arg);

		// pop it in the UI
		ui_popup_time(temp.len() / 40 + 2, "%s", temp.cstr());
	}
}


/*-------------------------------------------------
    logerror - log to the debugger and any other
    OSD-defined output streams
-------------------------------------------------*/

void CLIB_DECL logerror(const char *format, ...)
{
	if (global_machine != NULL)
	{
		va_list arg;
		va_start(arg, format);
		global_machine->vlogerror(format, arg);
		va_end(arg);
	}
}



/***************************************************************************
    INTERNAL INITIALIZATION LOGIC
***************************************************************************/

/*-------------------------------------------------
    mame_parse_ini_files - parse the relevant INI
    files and apply their options
-------------------------------------------------*/

void mame_parse_ini_files(core_options *options, const game_driver *driver)
{
	/* parse the INI file defined by the platform (e.g., "mame.ini") */
	/* we do this twice so that the first file can change the INI path */
	parse_ini_file(options, CONFIGNAME, OPTION_PRIORITY_MAME_INI);
	parse_ini_file(options, CONFIGNAME, OPTION_PRIORITY_MAME_INI);

	/* debug mode: parse "debug.ini" as well */
	if (options_get_bool(options, OPTION_DEBUG))
		parse_ini_file(options, "debug", OPTION_PRIORITY_DEBUG_INI);

	/* if we have a valid game driver, parse game-specific INI files */
	if (driver != NULL)
	{
#ifndef MESS
		const game_driver *parent = driver_get_clone(driver);
		const game_driver *gparent = (parent != NULL) ? driver_get_clone(parent) : NULL;
		machine_config *config;

		/* parse "vertical.ini" or "horizont.ini" */
		if (driver->flags & ORIENTATION_SWAP_XY)
			parse_ini_file(options, "vertical", OPTION_PRIORITY_ORIENTATION_INI);
		else
			parse_ini_file(options, "horizont", OPTION_PRIORITY_ORIENTATION_INI);

		/* parse "vector.ini" for vector games */
		config = global_alloc(machine_config(driver->machine_config));
		for (const screen_device_config *devconfig = screen_first(*config); devconfig != NULL; devconfig = screen_next(devconfig))
			if (devconfig->screen_type() == SCREEN_TYPE_VECTOR)
			{
				parse_ini_file(options, "vector", OPTION_PRIORITY_VECTOR_INI);
				break;
			}
		global_free(config);

		/* next parse "source/<sourcefile>.ini"; if that doesn't exist, try <sourcefile>.ini */
		astring sourcename;
		core_filename_extract_base(&sourcename, driver->source_file, TRUE)->ins(0, "source" PATH_SEPARATOR);
		if (!parse_ini_file(options, sourcename, OPTION_PRIORITY_SOURCE_INI))
		{
			core_filename_extract_base(&sourcename, driver->source_file, TRUE);
			parse_ini_file(options, sourcename, OPTION_PRIORITY_SOURCE_INI);
		}

		/* then parent the grandparent, parent, and game-specific INIs */
		if (gparent != NULL)
			parse_ini_file(options, gparent->name, OPTION_PRIORITY_GPARENT_INI);
		if (parent != NULL)
			parse_ini_file(options, parent->name, OPTION_PRIORITY_PARENT_INI);
#endif	/* MESS */
		parse_ini_file(options, driver->name, OPTION_PRIORITY_DRIVER_INI);
	}
}


/*-------------------------------------------------
    parse_ini_file - parse a single INI file
-------------------------------------------------*/

static int parse_ini_file(core_options *options, const char *name, int priority)
{
	file_error filerr;
	mame_file *file;

	/* don't parse if it has been disabled */
	if (!options_get_bool(options, OPTION_READCONFIG))
		return FALSE;

	/* open the file; if we fail, that's ok */
	astring fname(name, ".ini");
	filerr = mame_fopen_options(options, SEARCHPATH_INI, fname, OPEN_FLAG_READ, &file);
	if (filerr != FILERR_NONE)
		return FALSE;

	/* clear flag for added devices */
	options_set_bool(options, OPTION_ADDED_DEVICE_OPTIONS, FALSE, OPTION_PRIORITY_CMDLINE);
	
	/* update game name so depending callback options could be added */
	if (priority==OPTION_PRIORITY_DRIVER_INI) {		
		options_force_option_callback(options, OPTION_GAMENAME, name, priority);
	}
	
	/* parse the file and close it */
	options_parse_ini_file(options, mame_core_file(file), priority);
	mame_fclose(file);
	return TRUE;
}
